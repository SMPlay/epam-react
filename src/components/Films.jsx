import React from "react";
import { Link } from "react-router-dom";
import { createUseStyles } from "react-jss";
import FilmCard from "./FilmCard";
import Loading from "./Loading";

const useStyles = createUseStyles({
  films: {
    display: "grid",
    gridTemplateColumns: "repeat(3, 1fr)",
    gridGap: "100px",
    marginTop: 50,
  },
  default: {
    display: "flex",
    height: "100vh",
    justifyContent: "center",
    alignItems: "center",
  },
  filmsNotFoundTitle: {
    fontSize: 50,
    color: "rgb(128, 128, 128)",
  },
});

const Films = ({ films, isLoaded, getCurrentFilm }) => {
  const classes = useStyles();

  const content = films.map((film) => {
    const genres = film.genres.length === 1
      ? film.genres[0]
      : `${film.genres[0]} & ${film.genres[1]}`;
    return (
      <Link onClick={getCurrentFilm.bind(this, film.id)} to={`/movies/${film.id}`} key={film.id}>
        <FilmCard
          title={film.title}
          img={film.poster_path}
          genre={genres}
          dateRelease={film.release_date.slice(0, 4)}
        />
      </Link>
    );
  });
  const notFound = <p className={classes.filmsNotFoundTitle}>Films not found</p>;

  return (
    <main>
      <div className="container">
        <div className="row">
          <div className={(isLoaded && films.length !== 0) ? classes.films : classes.default}>
            {isLoaded ? content : <Loading />}
            {films.length === 0 && isLoaded && notFound}
          </div>
        </div>
      </div>
    </main>
  );
};

export default Films;
