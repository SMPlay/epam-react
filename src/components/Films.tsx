import React from "react";
import { Link } from "react-router-dom";
import { createUseStyles } from "react-jss";
import { FilmCard } from "./FilmCard";
import { Film } from "../types";

const useStyles = createUseStyles({
  films: {
    display: "grid",
    gridTemplateColumns: "repeat(3, 1fr)",
    gridGap: "100px",
    marginTop: "50px",
  },
  default: {
    display: "flex",
    height: "100vh",
    justifyContent: "center",
    alignItems: "center",
  },
  filmsNotFoundTitle: {
    fontSize: "50px",
    color: "rgb(128, 128, 128)",
  },
});

interface Props {
  films: Film[];
}

export const Films: React.FC<Props> = ({ films }) => {
  const classes = useStyles();

  const notFound = <p className={classes.filmsNotFoundTitle} data-test-notfound="not-found">Films not found</p>;
  return (
    <main>
      <div className="container">
        <div className="row">
          <div className={(films.length !== 0) ? classes.films : classes.default}>
            {films.map((film) => (
              <Link
                to={`/movies/${film.id}`}
                key={film.id}
                data-test-card={`film-${film.id}`}
              >
                <FilmCard
                  title={film.title}
                  imgUrl={film.poster_path}
                  genres={film.genres}
                  dateRelease={film.release_date.slice(0, 4)}
                />
              </Link>
            ))}
            {films.length === 0 && notFound}
          </div>
        </div>
      </div>
    </main>
  );
};
