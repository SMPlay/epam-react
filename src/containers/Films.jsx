import React, { useState, useEffect, useContext } from "react";
import { createUseStyles } from "react-jss";
import { Link } from "react-router-dom";
import Wrapper from "components/Wrapper";
import Film from "components/Film";
import NoFound from "components/Intermediate/NoFound";
import Loading from "components/Intermediate/Loading";
import PropTypes from "prop-types";
import FilmsContext from "../context/FilmsContext";

const useStyles = createUseStyles({
  allFilms: {
    marginTop: 50,
  },
  films: {
    display: "grid",
    gridTemplateColumns: "repeat(3, 1fr)",
    gridGap: "100px",
  },
  waitingOrLoaded: {
    height: "100vh",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  filmsNotFoundTitle: {
    fontSize: 50,
    color: "rgb(128, 128, 128)",
  },
});

const Films = ({ films }) => {
  const classes = useStyles();
  const [isLoaded, setIsLoaded] = useState(false);
  const { sortFilmsByGenre } = useContext(FilmsContext);

  useEffect(() => {
    if (films.length !== 0) {
      setIsLoaded(true);
    }
  }, [films]);

  let content;
  if (!isLoaded) {
    content = (
      <div className={classes.waitingOrLoaded}>
        <Loading />
      </div>
    );
  } else if (films.length === 0) {
    content = (
      <div className={classes.waitingOrLoaded}>
        <NoFound text="No film found" />
      </div>
    );
  } else {
    const filmsData = films.map((film) => {
      let genre = "";
      if (film.genres.length > 1) {
        genre = `${film.genres[0]} & ${film.genres[1]}`;
      } else {
        genre = `${film.genres[0]}`;
      }

      const dataRelease = film.release_date.slice(0, 4);
      return (
        <Link
          onClick={sortFilmsByGenre.bind(this, film.id)}
          key={film.id}
          to={`/movies/${film.id}`}
        >
          <Film
            img={film.poster_path}
            title={film.title}
            genre={genre}
            dateRelease={dataRelease}
          />
        </Link>
      );
    });

    content = <div className={classes.films}>{filmsData}</div>;
  }
  return (
    <main className={classes.allFilms}>
      <Wrapper padding={20}>{content}</Wrapper>
    </main>
  );
};

Films.defaultProps = {
  films: [],
};

Films.propTypes = {
  films: PropTypes.arrayOf(PropTypes.object),
};

export default Films;
