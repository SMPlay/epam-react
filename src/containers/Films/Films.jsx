import React, { useContext, useState, useEffect } from "react";
import { createUseStyles } from "react-jss";
import Wrapper from "components/Wrapper";
import Film from "components/Film";
import NoFound from "components/NoFound";
import Loading from "components/Loading";
import FilmsContext from "../../context/FilmsContext";

const useStyles = createUseStyles({
  films: {
    display: "grid",
    gridTemplateColumns: "repeat(3, 1fr)",
  },
  waitingOrLoaded: {
    height: "100vh",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  filmsNotFoundTitle: {
    fontSize: 50,
    color: "rgb(128, 128, 128)",
  },
});

const Films = () => {
  const classes = useStyles();
  const [isLoaded, setIsLoaded] = useState(false);
  const { foundFilms } = useContext(FilmsContext);

  useEffect(() => {
    if (foundFilms.length !== 0) {
      setIsLoaded(true);
    }
  }, [foundFilms]);

  let content;
  if (!isLoaded) {
    content = (
      <div className={classes.waitingOrLoaded}>
        <Loading />
      </div>
    );
  } else if (foundFilms.length === 0) {
    content = (
      <div className={classes.waitingOrLoaded}>
        <NoFound text="No film found" />
      </div>
    );
  } else {
    const filmsData = foundFilms.map((film) => {
      let genre = "";
      if (film.genres.length > 1) {
        genre = `${film.genres[0]} & ${film.genres[1]}`;
      } else {
        genre = `${film.genres[0]}`;
      }

      const dataRelease = film.release_date.slice(0, 4);
      return (
        <Film
          key={film.id}
          img={film.poster_path}
          title={film.title}
          genre={genre}
          dateRelease={dataRelease}
        />
      );
    });

    content = <div className={classes.films}>{filmsData}</div>;
  }
  return (
    <main>
      <Wrapper padding={20}>{content}</Wrapper>
    </main>
  );
};

export default Films;
