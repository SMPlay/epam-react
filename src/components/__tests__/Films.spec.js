import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Films from "../Films";

describe("Films component", () => {
  let props;
  beforeEach(() => {
    props = {
      getCurrentFilm: jest.fn(),
      isLoaded: true,
      films: [
        {
          id: 1,
          title: "Tomb Dider",
          genres: ["Actions", "Adventure"],
          vote_average: 8,
          release_date: "2020-02-10",
          poster_path: "",
        },
        {
          id: 2,
          title: "Transformers",
          genres: ["Robots", "Actions"],
          vote_average: 2,
          release_date: "2019-03-15",
          poster_path: "",
        },
        {
          id: 3,
          title: "Green line",
          genres: ["Drama"],
          vote_average: 3,
          release_date: "2019-05-30",
          poster_path: "",
        },
        {
          id: 4,
          title: "Zoopark",
          genres: ["Cartoons"],
          vote_average: 10,
          release_date: "2018-07-01",
          poster_path: "",
        },
      ],
    };
  });

  it("should render with state from props", () => {
    const tree = renderer.create(
      <Router>
        <Films
          films={props.films}
          isLoaded={props.isLoaded}
          getCurrentFilm={props.getCurrentFilm}
        />
      </Router>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render 4 films", () => {
    const component = mount(
      <Router>
        <Films
          films={props.films}
          isLoaded={props.isLoaded}
          getCurrentFilm={props.getCurrentFilm}
        />
      </Router>,
    );

    expect(component.find(".row a").length).toBe(4);
  });

  it("should render loading", () => {
    const component = mount(
      <Router>
        <Films
          films={props.films}
          isLoaded={!props.isLoaded}
          getCurrentFilm={props.getCurrentFilm}
        />
      </Router>,
    );

    expect(component.find(".row Loading").length).toBe(1);
  });

  it("should return not found", () => {
    const component = mount(
      <Router>
        <Films
          films={[]}
          isLoaded={props.isLoaded}
          getCurrentFilm={props.getCurrentFilm}
        />
      </Router>,
    );

    expect(component.find(".row p").text()).toBe("Films not found");
  });

  it("should return 1 genre", () => {
    const component = mount(
      <Router>
        <Films
          films={props.films}
          isLoaded={props.isLoaded}
          getCurrentFilm={props.getCurrentFilm}
        />
      </Router>,
    );

    expect(component.find(".row FilmCard").last().prop("genre")).toBe("Cartoons");
  });

  it("should return multiply genre", () => {
    const component = mount(
      <Router>
        <Films
          films={props.films}
          isLoaded={props.isLoaded}
          getCurrentFilm={props.getCurrentFilm}
        />
      </Router>,
    );

    expect(component.find(".row FilmCard").first().prop("genre")).toBe("Actions & Adventure");
  });
});
