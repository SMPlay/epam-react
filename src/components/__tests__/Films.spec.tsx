import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import { Films } from "../Films";

describe("Films component", () => {
  interface FakeFilm {
    id: number;
    title: string;
    genres: string[];
    vote_average: number;
    release_date: string;
    poster_path: string;
  }

  interface Props {
    films: FakeFilm[];
    onClick: () => void;
  }

  let props: Props;
  beforeEach(() => {
    props = {
      films: [
        {
          id: 1,
          title: "Tomb Dider",
          genres: ["Actions", "Adventure"],
          vote_average: 8,
          release_date: "2020-02-10",
          poster_path: "",
        },
        {
          id: 2,
          title: "Transformers",
          genres: ["Robots", "Actions"],
          vote_average: 2,
          release_date: "2019-03-15",
          poster_path: "",
        },
        {
          id: 3,
          title: "Green line",
          genres: ["Drama"],
          vote_average: 3,
          release_date: "2019-05-30",
          poster_path: "",
        },
        {
          id: 4,
          title: "Zoopark",
          genres: ["Cartoons"],
          vote_average: 10,
          release_date: "2018-07-01",
          poster_path: "",
        },
      ],
      onClick: jest.fn(),
    };
  });

  it("should render with state from props", () => {
    const tree = renderer.create(
      <Router>
        <Films
          films={props.films}
          onClick={props.onClick}
        />
      </Router>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render 4 films", () => {
    const component = mount(
      <Router>
        <Films
          films={props.films}
          onClick={props.onClick}
        />
      </Router>,
    );

    expect(component.find(".row a").length).toBe(4);
  });

  it("should return not found", () => {
    const component = mount(
      <Router>
        <Films
          films={[]}
          onClick={props.onClick}
        />
      </Router>,
    );

    expect(component.find(".row p").text()).toBe("Films not found");
  });

  it("should return 1 genre", () => {
    const component = mount(
      <Router>
        <Films
          films={props.films}
          onClick={props.onClick}
        />
      </Router>,
    );

    expect(component.find(".row FilmCard").last().find("p").text()).toBe("Cartoons");
  });

  it("should return multiply genre", () => {
    const component = mount(
      <Router>
        <Films
          films={props.films}
          onClick={props.onClick}
        />
      </Router>,
    );

    expect(component.find(".row FilmCard").first().find("p").text()).toBe("Actions & Adventure");
  });
});
