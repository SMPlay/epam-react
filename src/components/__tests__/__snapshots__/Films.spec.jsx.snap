// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Films component should render with state from props 1`] = `
<main>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="films-0-2-1"
      >
        <a
          data-test-card="film-1"
          href="/movies/1"
          onClick={[Function]}
        >
          <div
            className="film-0-2-4"
          >
            <figure>
              <img
                alt="Tomb Dider2020"
                className="filmImg-0-2-5"
                src=""
              />
              <figcaption
                className="filmDescription-0-2-6"
              >
                <div>
                  <h4
                    className="title-0-2-8"
                  >
                    Tomb Dider
                  </h4>
                  <p
                    className="genre-0-2-9"
                  >
                    Adventure & Adventure
                  </p>
                </div>
                <div
                  className="dateRelease-0-2-7"
                >
                  2020
                </div>
              </figcaption>
            </figure>
          </div>
        </a>
        <a
          data-test-card="film-2"
          href="/movies/2"
          onClick={[Function]}
        >
          <div
            className="film-0-2-4"
          >
            <figure>
              <img
                alt="Transformers2019"
                className="filmImg-0-2-5"
                src=""
              />
              <figcaption
                className="filmDescription-0-2-6"
              >
                <div>
                  <h4
                    className="title-0-2-8"
                  >
                    Transformers
                  </h4>
                  <p
                    className="genre-0-2-9"
                  >
                    Actions & Actions
                  </p>
                </div>
                <div
                  className="dateRelease-0-2-7"
                >
                  2019
                </div>
              </figcaption>
            </figure>
          </div>
        </a>
        <a
          data-test-card="film-3"
          href="/movies/3"
          onClick={[Function]}
        >
          <div
            className="film-0-2-4"
          >
            <figure>
              <img
                alt="Green line2019"
                className="filmImg-0-2-5"
                src=""
              />
              <figcaption
                className="filmDescription-0-2-6"
              >
                <div>
                  <h4
                    className="title-0-2-8"
                  >
                    Green line
                  </h4>
                  <p
                    className="genre-0-2-9"
                  >
                    Drama
                  </p>
                </div>
                <div
                  className="dateRelease-0-2-7"
                >
                  2019
                </div>
              </figcaption>
            </figure>
          </div>
        </a>
        <a
          data-test-card="film-4"
          href="/movies/4"
          onClick={[Function]}
        >
          <div
            className="film-0-2-4"
          >
            <figure>
              <img
                alt="Zoopark2018"
                className="filmImg-0-2-5"
                src=""
              />
              <figcaption
                className="filmDescription-0-2-6"
              >
                <div>
                  <h4
                    className="title-0-2-8"
                  >
                    Zoopark
                  </h4>
                  <p
                    className="genre-0-2-9"
                  >
                    Cartoons
                  </p>
                </div>
                <div
                  className="dateRelease-0-2-7"
                >
                  2018
                </div>
              </figcaption>
            </figure>
          </div>
        </a>
      </div>
    </div>
  </div>
</main>
`;
